File: heap/k_largest_elements_immutable_max_heap.c
/*
 * author: greyshell
 * description: find kth largest elements in a immutable max heap
 * */

#include <stdio.h>
#include "stdbool.h"
#include "../libO2/include/binary_heap_dynamic_array_int.h"


void kth_largest_elements_immutable_max_heap(heap *h_immutable, int n, int k, int *out_arr) {
    /*
     * time complexity:
     * space complexity:
     */
    heap h;
    int out_data;
    size_t top_idx;

    // create a max heap
    initialize_heap(&h, k, true);

    // Put top element in heap
    push_heap(&h, 0);

    for (int i = 0;i<k;i++) {
        pop_heap(&h, (int *)&top_idx);
        peek_heap_by_index(&h, top_idx, &out_arr[i]);

        size_t left_child = get_heap_node_left_child_index(top_idx);
        size_t right_child = get_heap_node_right_child_index(top_idx);
        push_heap(&h, left_child);
        push_heap(&h, right_child);
    }


    delete_heap(&h);
}

/* 17 7 16 2 3 15 14 <- h_immutable

 Init: 17

 1'st: 16 7 [17]

 2'nd: 15 7 14 [17, 16]

 3rd: 14 7 [17, 16, 15]

 4th: 7 [17, 16, 15, 14]

*/


int main(void) {
    // test the solution
    int test_arr[] = {7, 17, 16, 2, 3, 15, 14};
    int n = 7;
    int k = 4;
    int out_arr[4] = {0};
    heap h;

    build_heap(&h, true, test_arr, 7);
    display_heap(&h);


    kth_largest_elements_immutable_max_heap(&h, n, k, out_arr);

    // expected output: 17, 16, 15, 14
//    for (int i = 0; i < k; i++){
//        printf("%d ", out_arr[i]);
//    }
//    printf("\n");

    return 0;
}

File: heap/k_largest_elements_immutable_max_heap.py
#!/usr/bin/env python3

# author: greyshell
# description: find k largest elements from a immutable max heap

from snowowl import Heap, HeapType


class Node:
    def __init__(self, value, index):
        self.value = value
        self.index = index

    def __lt__(self, other):
        return self.value < other.value

    def __eq__(self, other):
        return self.value == other.value


def get_k_largest_elements_immutable_max_heap(immutable_max_heap: Heap, k: int) -> list:
    """
    time complexity: O(k * log k)
    space complexity: O(k) -> auxiliary max heap
    """
    # create an auxiliary max heap
    auxiliary_max_heap = Heap([], heap_type=HeapType.MAX)

    # peek the min item from the immutable max heap
    # create a node obj with the value and index and push that object into auxiliary max heap
    node = Node(immutable_max_heap.peek(), 0)
    auxiliary_max_heap.insert(node)

    result = list()
    for i in range(0, k):
        node = auxiliary_max_heap.remove()
        result.append(node.value)

        index = node.index

        left_child_index = 2 * index + 1

        if left_child_index < len(immutable_max_heap):
            left_child = immutable_max_heap.__getitem__(left_child_index)
            left_node = Node(left_child, left_child_index)
            auxiliary_max_heap.insert(left_node)

        right_child_index = 2 * index + 2

        if right_child_index < len(immutable_max_heap):
            right_child = immutable_max_heap.__getitem__(right_child_index)
            right_node = Node(right_child, right_child_index)
            auxiliary_max_heap.insert(right_node)

    return result
